src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -967,6 +971,20 @@ void primitive_inst::realloc_if_needed(bool prev_execution_skipped) {
                                           is_output_buffer(this, true),
                                           output_memory_ptr(i).get(),
                                           true);
+            // memory_pool does not reset padded memory to be reused.
+            if (static_cast<bool>(_outputs[i]->get_layout().data_padding)
+                && need_reset_output_memory() && !can_be_optimized()) {
+                auto mem_ptr = _outputs[i];
+                auto num_users = memory_pool::count_users_in_padded_pool(_network.get_memory_pool(), mem_ptr);
+                const auto& updated_shape = updated_layouts[i].get_shape();
+                const auto& params_shape = updated_params.output_layouts[i].get_shape();
+                // It need to reset if the memory is reused in a flatten shape.
+                if (num_users > 1 && (updated_shape != params_shape) && is_flattened(params_shape)) {
+                    GPU_DEBUG_TRACE_DETAIL << id() << " : Need reset padded output memory considering user" << std::endl;
+                    add_dep_event(_outputs[i]->fill(_network.get_stream()));
+                }
+            }
```

Comment:
* is it OK for non-padded case? From `get_from_non_padded_pool` function, I'm afraid the same issue may happen for non-padded case.
* I think such functionality should be implemented within allocate_output function with additional reset flag. This is conflicting with existing flag.